docs/configuration.md
## Configuration

### Credentials

The ECS Local Endpoints container uses the AWS SDK for Go, and thus it supports all of its [methods of configuration](https://docs.aws.amazon.com/sdk-for-go/v1/developer-guide/configuring-sdk.html). We recommend providing credentials via an AWS CLI Profile. To do this, mount `$HOME/.aws/` ([`%UserProfile%\.aws` on Windows](https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-files.html)) into the container. As shown in the example Compose file, the container path of the volume should be `/home/.aws/` because the environment variable `HOME` is set to `/home` in the image. This way, inside the container, the SDK will be able to find credentials at `$HOME/.aws/`. To use a non-default profile, set the `AWS_PROFILE` environment variable on the Local Endpoints container.

The Local Endpoints container will retrieve temporary session credentials from STS.  To provide a custom CA bundle for the STS client, mount your certificates file into the Local Endpoints container at any of the following locations:
* `/etc/ssl/certs/ca-certificates.crt`
* `/etc/pki/tls/certs/ca-bundle.crt`
* `/etc/ssl/ca-bundle.pem`
* `/etc/pki/tls/cacert.pem`
* `/etc/pki/ca-trust/extracted/pem/tls-ca-bundle.pem`

For example, on an Ubuntu machine, you can mount your machine's certificates file at `/etc/ssl/certs/ca-certificates.crt` into the Local Endpoint container at `/etc/ssl/certs/ca-certificates.crt`.

### Custom IAM and STS Endpoints

Local Endpoints can be configured to use custom IAM and STS endpoints. Simply define the `IAM_ENDPOINT` and `STS_ENDPOINT` environment variables in the Local Endpoints container.

This may be useful in scenarios where your application container is configured to obtain credentials from ECS (see [Vend Credentials to Containers](features.md#vend-credentials-to-containers)), but you do not want to provide Local Endpoints with AWS credentials. Providing an IAM and STS simulator and configuring the Local Endpoints container with custom IAM and STS endpoints enables testing without an AWS account.

See [`docker-compose.localstack.yml`](../examples/docker-compose.localstack.yml) for an example.

### Docker

Local Endpoints responds to Metadata requests with real data about the containers running on your machine. In order to do this, you must mount the [Docker socket](https://docs.docker.com/engine/reference/commandline/dockerd/#daemon-socket-option) into the container. Make sure the Local Endpoints container is given a volume with source path `/var/run` and container path `/var/run`.

### Environment Variables

General Configuration:
* `ECS_LOCAL_METADATA_PORT` - Set the port that the container listens at. The default is `80`.
* `IAM_ENDPOINT` - Set the endpoint used by the AWS SDK for IAM. The default is undefined, which results in using the default AWS region.
* `STS_ENDPOINT` - Set the endpoint used by the AWS SDK for STS. The default is undefined, which results in using the default AWS region.

Task Metadata Configuration: while Local Endpoints returns real runtime information obtained from Docker in metadata requests, some values have no relevance locally and are mocked:
* `CLUSTER_ARN` - Set the 'cluster' name which is returned in Task Metadata responses. Default: `ecs-local-cluster`.
* `TASK_ARN` - Set ARN of the mock local 'task' which your containers will appear to be part of in Task Metadata responses. Default: `arn:aws:ecs:us-west-2:111111111111:task/ecs-local-cluster/37e873f6-37b4-42a7-af47-eac7275c6152`.
* `TASK_DEFINITION_FAMILY` - Set family name for the mock task definition which your containers will appear to be part of in Task Metadata responses. Default: `esc-local-task-definition`.
* `TASK_DEFINITION_REVISION` - Set the Task Definition revision. Default: `1`.

Credentials Configuration:
* `SHARED_TOKEN_EXPIRATION` - Set an expiration duration (quantity + unit) for shared credentials when a session token is provided. This provides a hint for clients to refresh their credentials periodically. The default is 750s (12.5 minutes), which results in some clients (notably Boto3) opportunistically refreshing credentials in a background thread.


CHANGELOG.md
# Changelog

## 1.4.0
* Feature - Add support for ARM64 binaries and docker images (#59)

## 1.3.0
* Feature - Add support for V4 and generic metadata endpoints (#38)

## 1.2.0
* Feature - Add support for assuming roles in other accounts with path `/role-arn/{role arn}` (#36)

## 1.1.0
* Bug - Set expiration timestamp on temporary credentials (#26)
* Feature - Change base image to amazonlinux to support sourcing credentials from an external process (#30)
* Feature - Add support for custom endpoints for STS and IAM (#16)
* Enhancement - Print verbose error messages for credential chain problems (#25)

## 1.0.1
* Enhancement - Add custom user agent header for calls made to STS and IAM (#9)

## 1.0.0
* Feature - Support vending temporary credentials to containers from a base set of credentials
* Feature - Support vending temporary credentials to containers from an IAM Role
* Feature - [Support Task Metadata V2 Paths](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task-metadata-endpoint-v2.html):
	- Task Metadata - `/v2/metadata`
	- Container Metadata - `/v2/metadata/<container-id>`
	- Task Stats - `/v2/stats`
	- Container Stats - `/v2/stats/<container-id>`
* Feature - [Support Task Metadata V3 Paths](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task-metadata-endpoint-v3.html):
	- Container Metadata - `/v3` OR `/v3/containers/<container>`
	- Task Metadata - `/v3/task` OR `/v3/containers/<container>/task`
	- Container Stats - `/v3/stats` OR `/v3/containers/<container>/stats`
	- Task Stats - `/v3/task/stats` OR `/v3/containers/<container>/task/stats`


local-container-endpoints/config/config.go
// Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License"). You may
// not use this file except in compliance with the License. A copy of the
// License is located at
//
//	http://aws.amazon.com/apache2.0/
//
// or in the "license" file accompanying this file. This file is distributed
// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
// express or implied. See the License for the specific language governing
// permissions and limitations under the License.

// Package config contains environment variables and default values for Local Endpoints
package config

// Environment Variables
const (
	// PortVar defines the port that metadata and credentials listen at
	PortVar = "ECS_LOCAL_METADATA_PORT"

	// Metadata related
	ClusterARNVar            = "CLUSTER_ARN"
	TaskARNVar               = "TASK_ARN"
	TDFamilyVar              = "TASK_DEFINITION_FAMILY"
	TDRevisionVar            = "TASK_DEFINITION_REVISION"
	ContainerInstanceTagsVar = "CONTAINER_INSTANCE_TAGS"
	TaskTagsVar              = "TASK_TAGS_VAR"

	// Custom endpoint related
	IAMCustomEndpointVar = "IAM_ENDPOINT"
	STSCustomEndpointVar = "STS_ENDPOINT"

	// Shared credentials default expiration value when a token is detected.
	SharedTokenExpirationVar = "SHARED_TOKEN_EXPIRATION"

	// User-defined, static metadata that overrides/augments the normal response
	ContainerMetadataPathVar = "CONTAINER_METADATA_PATH"
	TaskMetadataPathVar      = "TASK_METADATA_PATH"
)

// Defaults
const (
	// DefaultPort is the default port the server listens at
	DefaultPort = "80"

	// Metadata related
	DefaultContainerType = "NORMAL"
	DefaultClusterName   = "ecs-local-cluster"
	DefaultTaskARN       = "arn:aws:ecs:us-west-2:111111111111:task/ecs-local-cluster/37e873f6-37b4-42a7-af47-eac7275c6152"
	DefaultTDFamily      = "esc-local-task-definition"
	DefaultTDRevision    = "1"

	// Expire shared credentials with a token in 12.5 minutes.
	DefaultSharedTokenExpiration = 750
)

// Settings
const (
	HTTPTimeoutDuration = "5s"
)

// URL Paths

// Credentials
const (
	// RoleCredentialsPath is the path for obtaining credentials from a role
	RoleCredentialsPath = "/role/{role}"
	// RoleCredentialsPathWithSlash adds a trailing slash
	RoleCredentialsPathWithSlash = RoleCredentialsPath + "/"

	//RoleArnCredentialsPath is the path for obtaining credentials from a role ARN
	RoleArnCredentialsPath = "/role-arn/{roleArn}/{roleName}"
	// RoleArnCredentialsPathWithSlash adds a trailing slash
	RoleArnCredentialsPathWithSlash = RoleArnCredentialsPath + "/"

	// TempCredentialsPath is the path for obtaining temp creds from sts:GetSessionsToken
	TempCredentialsPath = "/creds"
	// TempCredentialsPathWithSlash adds a trailing slash
	TempCredentialsPathWithSlash = TempCredentialsPath + "/"
)

// V3
const (
	// V3ContainerMetadataPath is the path for V3 container metadata
	V3ContainerMetadataPath = "/v3"
	// V3ContainerMetadataPathWithSlash adds a trailing slash
	V3ContainerMetadataPathWithSlash = V3ContainerMetadataPath + "/"
	// V3ContainerMetadataPathWithIdentifier is the V3 container metadata path with an identifer specified
	V3ContainerMetadataPathWithIdentifier = "/v3/containers/{identifier}"
	// V3ContainerMetadataPathWithIdentifierAndSlash adds a trailing slash
	V3ContainerMetadataPathWithIdentifierAndSlash = V3ContainerMetadataPathWithIdentifier + "/"

	// V3ContainerStatsPath is the path for V3 container stats
	V3ContainerStatsPath = "/v3/stats"
	// V3ContainerStatsPathWithSlash adds a trailing slash
	V3ContainerStatsPathWithSlash = V3ContainerStatsPath + "/"
	// V3ContainerStatsPathWithIdentifier is the V3 container stats path with an identifier
	V3ContainerStatsPathWithIdentifier = "/v3/containers/{identifier}/stats"
	// V3ContainerStatsPathWithIdentifierAndSlash adds a trailing slash
	V3ContainerStatsPathWithIdentifierAndSlash = V3ContainerStatsPathWithIdentifier + "/"

	// V3TaskMetadataPath is the path for V3 task metadata
	V3TaskMetadataPath = "/v3/task"
	// V3TaskMetadataPathWithSlash adds a trailing slash
	V3TaskMetadataPathWithSlash = V3TaskMetadataPath + "/"
	// V3TaskMetadataPathWithIdentifier is the v3 task metadata path with an identifier
	V3TaskMetadataPathWithIdentifier = "/v3/containers/{identifier}/task"
	// V3TaskMetadataPathWithIdentifierWithSlash adds a trailing slash
	V3TaskMetadataPathWithIdentifierWithSlash = V3TaskMetadataPathWithIdentifier + "/"

	// V3TaskStatsPath is the path for V3 task stats
	V3TaskStatsPath = "/v3/task/stats"
	// V3TaskStatsPathWithSlash adds a trailing slash
	V3TaskStatsPathWithSlash = V3TaskStatsPath + "/"
	// V3TaskStatsPathWithIdentifier is the v3 task stats path with an identifier
	V3TaskStatsPathWithIdentifier = "/v3/containers/{identifier}/task/stats"
	// V3TaskStatsPathWithIdentifierAndSlash adds a trailing slash
	V3TaskStatsPathWithIdentifierAndSlash = V3TaskStatsPathWithIdentifier + "/"
)

// V2
const (
	// V2TaskMetadataPath is the V2 Task Metadata path
	V2TaskMetadataPath = "/v2/metadata"
	// V2TaskMetadataPathWithSlash adds a trailing slash
	V2TaskMetadataPathWithSlash = V2TaskMetadataPath + "/"

	// V2ContainerMetadataPath is the V2 Container Metadata path
	V2ContainerMetadataPath = "/v2/metadata/{identifier}"
	// V2ContainerMetadataPathWithSlash adds a trailing slash
	V2ContainerMetadataPathWithSlash = V2ContainerMetadataPath + "/"

	// V2TaskStatsPath is the V2 Task Stats Path
	V2TaskStatsPath = "/v2/stats"
	// V2TaskStatsPathWithSlash adds a trailing slash
	V2TaskStatsPathWithSlash = V2TaskStatsPath + "/"

	// V2ContainerStatsPath is the V2 container stats path
	V2ContainerStatsPath = "/v2/stats/{identifier}"
	// V2ContainerStatsPathWithSlash adds a trailing slash
	V2ContainerStatsPathWithSlash = V2ContainerStatsPath + "/"
)
